File: 03/part2.py
parts = []
lines = []
with open("sample3") as input:
    for line in input:
        lines.append(line.strip())

ratios = []
for i in range(0, len(lines)):  # for each line
    for j in range(0, len(lines[i])):  # for each char on line
        gears = {}

        if lines[i][j] == "*":
            # scan left
            if j > 0 and lines[i][j - 1].isdigit():
                parsed_number = ""
                k = j - 1
                while k >= 0 and lines[i][k].isdigit():
                    parsed_number = lines[i][k] + parsed_number
                    k -= 1
                k = max(0, k)
                if parsed_number != "":
                    gears.setdefault((i,k), [])
                    gears[(i,k)].append(int(parsed_number))

            # scan right
            if j < len(lines[i]) - 1 and lines[i][j + 1].isdigit():
                parsed_number = ""
                k = j + 1
                while k < len(lines[i]) and lines[i][k].isdigit():
                    parsed_number += lines[i][k]
                    k += 1
                k = min(len(lines[i]), k)
                if parsed_number != "":
                    gears.setdefault((i,k), [])
                    gears[(i,k)].append(int(parsed_number))

            # scan up
            if i > 0:
                parsed_number = ""
                index = 0
                for k in range(j - 1, j + 2):
                    if parsed_number != "" and not lines[i - 1][k].isdigit():
                        parsed_number = ""
                        continue
                    if k < len(lines[i - 1]) and lines[i - 1][k].isdigit():
                        # expand to the left
                        while k > 0 and lines[i - 1][k - 1].isdigit():
                            k -= 1
                        index = k

                    # scan to the right
                    while k < len(lines[i]) and lines[i - 1][k].isdigit():
                        parsed_number += lines[i - 1][k]
                        # if i == 10:
                        #     print(f"{parsed_number=}, {k=}")
                        if k < len(lines[i]):
                            k += 1

                    if parsed_number != "":
                        gears.setdefault((i-1,index), [])
                        gears[(i-1,index)].append(int(parsed_number))
                        # if i == 10:
                        #     print(f"adding {parsed_number} to {(i-1,index)}, {i=}, {index=}, {k=}")

            # scan down
            if i < len(lines) - 1:
                parsed_number = ""
                index = 0
                for k in range(j - 1, j + 2):
                    print(f"{parsed_number=}, {k=}")
                    if parsed_number != "" and not lines[i + 1][k].isdigit():
                        parsed_number = ""
                        continue
                    if k < len(lines[i + 1]) and lines[i + 1][k].isdigit():
                        # expand to the left
                        while k > 0 and lines[i + 1][k - 1].isdigit():
                            k -= 1
                        index = k

                    # scan to the right
                    while k < len(lines[i]) and lines[i + 1][k].isdigit():
                        parsed_number += lines[i + 1][k]
                        if k < len(lines[i]):
                            k += 1

                if parsed_number != "":
                    gears.setdefault((i+1,index), [])
                    gears[(i+1,index)].append(int(parsed_number))

        # remove empty arrays
        gears = {k: v for k, v in gears.items() if v is not None}
        if len(gears) == 2:
            for key, value in gears.items():
                parts.append(value[0])
        elif len(gears) > 0:
            print(f"{gears} is not a valid gear...?")

# correct answer is 6756 (for sample2)
for i in range(0, len(parts)):
    if i % 2 > 0:
        print(f"multiplying {parts[i]} and {parts[i - 1]}")
        ratios.append(parts[i] * parts[i - 1])

# sum up all the ratios
sum = 0
for ratio in ratios:
    sum += ratio

# guesses: 
# - 63913958 (too low)
print("the answer is", sum)

File: 03/part2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"unicode"
)

type Part struct {
	number   int
	position int
}

type ParsedLine struct {
	parts   []Part
	symbols []int
}

type Gear struct {
	part1, part2 Part
}

func FindAdjacentParts(position int, parts []Part) []Part {
	adjacents := make([]Part, 0)
	for _, p := range parts {
		digits := strconv.Itoa(p.number)

		for i := range digits {
			if p.position+i == position || p.position+i == position-1 || p.position+i == position+1 {
				adjacents = append(adjacents, p)
				break
			}
		}
	}

	return adjacents
}

func FindGears(input []ParsedLine) []Gear {
	gears := make([]Gear, 0)
	for i := range input {
		for _, symbol := range input[i].symbols {

			adjacents := FindAdjacentParts(symbol, input[i].parts)
			if i > 0 {
				adjacents = append(adjacents, FindAdjacentParts(symbol, input[i-1].parts)...)
			}
			if i < len(input)-1 {
				adjacents = append(adjacents, FindAdjacentParts(symbol, input[i+1].parts)...)
			}

			if len(adjacents) == 2 {
				gears = append(gears, Gear{adjacents[0], adjacents[1]})
			}
		}
	}
	return gears
}

func ReadFile(path string) []string {
	lines := make([]string, 0)

	readFile, err := os.Open(path)
	if err != nil {
		panic(err)
	}

	defer readFile.Close()

	fileScanner := bufio.NewScanner(readFile)
	fileScanner.Split(bufio.ScanLines)

	for fileScanner.Scan() {
		lines = append(lines, fileScanner.Text())
	}
	return lines
}

func ParseInput(lines []string) []ParsedLine {
	data := make([]ParsedLine, 0)
	for _, line := range lines {
		pl := ParsedLine{
			parts:   FindPartNumbers(line),
			symbols: FindSymbols(line),
		}
		data = append(data, pl)
	}
	return data
}

func main() {
	data := ParseInput(ReadFile("input"))
	gears := FindGears(data)
	answer := 0
	for _, gear := range gears {
		answer += gear.part1.number * gear.part2.number
	}
	// Correct answer is 80703636
	fmt.Printf("The answer is %d\n", answer)
}

func FindSymbols(s string) []int {
	result := make([]int, 0)
	for i, r := range s {
		if r == '*' {
			result = append(result, i)
		}
	}
	return result
}

func FindPartNumbers(s string) []Part {
	matches := []Part{}

	parsed_number := ""
	number := 0

	for i := 0; i < len(s); i++ {
		if unicode.IsDigit([]rune(s)[i]) {
			parsed_number = fmt.Sprintf("%s%s", parsed_number, string([]rune(s)[i]))
			continue
		}
		if len(parsed_number) > 0 {
			var err error
			number, err = strconv.Atoi(parsed_number)
			if err != nil {
				panic(err)
			}
			matches = append(matches, Part{
				position: i - len(parsed_number),
				number:   number,
			})
			parsed_number = ""
		}
	}

	if len(parsed_number) > 0 {
		var err error
		number, err = strconv.Atoi(parsed_number)
		if err != nil {
			panic(err)
		}
		matches = append(matches, Part{
			position: len(s) - len(parsed_number),
			number:   number,
		})
		parsed_number = ""

	}

	return matches
}

func CalculateRatios(gears []Gear) []int {
	ratios := make([]int, len(gears))
	for _, gear := range gears {
		ratios = append(ratios, gear.part1.number*gear.part2.number)
	}
	return ratios
}
